### ISSUE TITLE
ESM-style plugin import doesn't pass TypeScript type check.

### ISSUE THREAD

**Issue by hongyiweiwu** [fastify/fastify#2403]
## To Reproduce

Pick any Fastify plugin (I'll use `fastify-formbody` for now):
```js
fastify().register(import('fastify-formbody'));
```

TypeScript compiler gives the following error:
```
No overload matches this call.
  Overload 1 of 3, '(plugin: FastifyPluginCallback<FastifyPluginOptions>, opts?: FastifyRegisterOptions<FastifyPluginOptions>): FastifyInstance<...> & PromiseLike<...>', gave the following error.
    Argument of type 'Promise<typeof import(".../node_modules/fastify-formbody/formbody")>' is not assignable to parameter of type 'FastifyPluginCallback<FastifyPluginOptions>'.
      Type 'Promise<typeof import(".../node_modules/fastify-formbody/formbody")>' provides no match for the signature '(instance: FastifyInstance<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse, FastifyLoggerInstance>, opts: FastifyPluginOptions, next: (err?: Error) => void): void'.
  Overload 2 of 3, '(plugin: FastifyPluginAsync<FastifyPluginOptions>, opts?: FastifyRegisterOptions<FastifyPluginOptions>): FastifyInstance<...> & PromiseLike<...>', gave the following error.
    Argument of type 'Promise<typeof import(".../node_modules/fastify-formbody/formbody")>' is not assignable to parameter of type 'FastifyPluginAsync<FastifyPluginOptions>'.
      Type 'Promise<typeof import(".../node_modules/fastify-formbody/formbody")>' provides no match for the signature '(instance: FastifyInstance<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse, FastifyLoggerInstance>, opts: FastifyPluginOptions): Promise<...>'.
  Overload 3 of 3, '(plugin: FastifyPluginCallback<FastifyPluginOptions> | FastifyPluginAsync<FastifyPluginOptions>, opts?: FastifyRegisterOptions<...>): FastifyInstance<...> & PromiseLike<...>', gave the following error.
    Argument of type 'Promise<typeof import(".../node_modules/fastify-formbody/formbody")>' is not assignable to parameter of type 'FastifyPluginCallback<FastifyPluginOptions> | FastifyPluginAsync<FastifyPluginOptions>'.
      Type 'Promise<typeof import(".../node_modules/fastify-formbody/formbody")>' is not assignable to type 'FastifyPluginAsync<FastifyPluginOptions>'.
```

## Expected behavior

There should be no errors.

We can use `as any` to bypass this problem but not sure it's the best thing to do.

## Your Environment

- *node version*: 14
- *fastify version*: 3.0.1
- *os*: Mac

**Comment by mcollina** [comment]
Interesting one! What’s the type of `import()?`? I don’t think Typescript can detect this automatically.

**Comment by mcollina** [comment]
cc @Ethan-Arrowood

**Comment by hongyiweiwu** [comment]
It's a special import type, and there doesn't seem to be a way to type what's actually imported.

This really goes back to the argument of whether to support es6 modules. I understand that Fastify doesn't yet, but it allows mimicking behaviors of es6 modules by setting `fastify.default = fastify` in the javascript file and `export default Fastify` in the typing file, so in Typescript I can just do `import fastify from 'fastify'` with `fastify` correctly typed. For most of the plugins though the plugin functions are default exported in the type declaration file, but in the JS implementation the `default` property isn't set. So the only way I can import these plugins is to do a dynamic import (in the example code), and TypeScript can't figure out the type of the imported function.

It'd be nice if there's some official guidelines on how to structure exports for plugins. Specifically do the same thing that the main `Fastify` package is doing, so I can use es6 style default import, which is correctly typed by TypeScript.

**Comment by mcollina** [comment]
> This really goes back to the argument of whether to support es6 modules. I understand that Fastify doesn't yet

This is incorrect as it totally does, there are even unit tests that verifies that. The fact that you can `app.register(import('foobar'))` was added to support native esm. Fastify does not support _faux modules_, which is the transpiled output of TS and babel. 

This is only a typescript issue.

**Comment by hongyiweiwu** [comment]
> This is incorrect as it totally does, there are even unit tests that verifies that

Sorry about this -- since I only use typescript, I only saw how these modules behave on typescript compiler outputs. When I do 
```typescript
import plugin from 'fastify-plugin'
app.register(plugin);
```

TypeScript transpiles it to
```typescript
const plugin = require('fastify-plugin')
app.register(plugin.default);
```

Because most of the plugins haven't manually set the `default` property that point to itself while having declared `export default` in the type declaration files, this problem can't be caught during compile time. During runtime we will see `plugin` is actually `undefined`.

As for the problem above, we can certainly type the `register` function to just take in `any`, but that's not ideal. A better solution might be having all plugins manually set the `default` property just like the `fastify` package, so we can use the default import syntax and get type safety.

**Comment by mcollina** [comment]
I might also be ok to support _faux modules_, i.e. look for the default property if it’s there.

**Comment by Ethan-Arrowood** [comment]
I'm confused about what issue is happening here.

If `import plugin from 'some-fastify-plugin'` works as expected why does it break when you do `import
('some-fastify-plugin')` inline? Is `import()` different than `import from`?

**Comment by hongyiweiwu** [comment]
> I'm confused about what issue is happening here.
> 
> If `import plugin from 'some-fastify-plugin'` works as expected why does it break when you do `import ('some-fastify-plugin')` inline? Is `import()` different than `import from`?

Let's say in the `.d.ts` file of the plugin you have `export default plugin` where `plugin` is of type `FastifyPlugin`. Then when you do `import plugin from 'plugin'` the variable `plugin` has type `FastifyPlugin`. On the other hand if you do `const a = import('plugin')` then `a` will have a special import type `typeof import('plugin')` which isn't equal to any of the normal types.

So it would seem that using `import plugin from 'plugin'` would solve everything, except it doesn't. Most Fastify plugins `export default` the plugin function directly in the `.d.ts` file, but in the corresponding `.js` file they use CommonJS style export. Now in TypeScript, when you do `import plugin from 'plugin'` and later refers to `plugin`, TypeScript transpiles `plugin` to `plugin.default`. But since the actual exposed object is the plugin function, it doesn't have the `default` property, so all such references are gonna be evaluated to `undefined`. Hence by current implementations of most plugins, we can't use this kind of import method. (Basically most plugins' export objects are incorrectly typed in the `d.ts` file)

This is pretty confusing tbh, but I think it boils down to two things
* fix the signature of `fastify.register` so it also takes in the special import type. This allows using `import()` style import while passing type checks.
* provide guidelines for plugins on how to properly type the export objects. This will allow `import plugin from 'plugin'` type of import without the `undefined` issue.

**Comment by Ethan-Arrowood** [comment]
I honestly wish I knew more about whats happening here but this is really melting my brain. 

Is this comment related? https://github.com/fastify/fastify-swagger/pull/264#issuecomment-656924808

>  (Basically most plugins' export objects are incorrectly typed in the d.ts file) 

I'd really like it if someone can lead up how to fix this. Plugins was (and still appears to be) the hardest part of the whole v3 types. 

> This is pretty confusing tbh, but I think it boils down to two things

Are these two things options? Or items apart of the same solution? In either case would you like to send a PR with the intended fix?

**Comment by hongyiweiwu** [comment]
I wish I'm a better explainer... But this entire import/export system is quite hopeless.

I'm not an expert on such things but this might be a way to solve the problem. We put in the official doc that plugins should be exported in this way:
```javascript
// plugin.js or plugin.ts if plugin is directly written in typescript.
const myPlugin = (fastify, opts, next) => {
   // Plugin stuff.
  next();
};

myPlugin.default = myPlugin;
module.exports = myPlugin;
```

```typescript
// plugin.d.ts only if the plugin is written in javascript
declare const plugin: FastifyPlugin;
export default plugin;
```

Now when I want to import the plugin:
```typescript
import plugin from 'plugin';  // plugin variable here has appropriate FastifyPlugin type.

app.register(plugin); // passes typescript typecheck, doesn't become undefined in runtime.
```

**Comment by Ethan-Arrowood** [comment]
> Does FastifyOptions sound good (since it will be the options object passed into the fastify constructor).

You are a very good explainer. Like I said this is really hard and I appreciate you sharing this information with us. It is very valuable and helpful!

---

What about plugins defined with `fastify-plugin`? Does all of the above processes still work?

**Comment by mcollina** [comment]
Would it help if `app.register({ default })` would load `default` as the plugin?

**Comment by hongyiweiwu** [comment]
> You are a very good explainer. Like I said this is really hard and I appreciate you sharing this information with us. It is very valuable and helpful!

Always a pleasure!

> What about plugins defined with `fastify-plugin`? Does all of the above processes still work?

It should work. The process described above works for all function exports.

> Would it help if `app.register({ default })` would load `default` as the plugin?

I don't think so. The problem is that the plugin authors are not exporting the default property, not that the plugin consumers are not using this property.

**Comment by mcollina** [comment]
I think a PR to `fastify-plugin` could alleviate most of the problems and set the `default` property on the function.
Also some docs PR to explain this.

**Comment by fox1t** [comment]
Hi to all. I really want to fix all of the issues that are popping out on the TS side. Actually we have a simple solution for all of our problems and it is not really that hard to support all of the methods that TS/ES2015/CJS have and we don't have to find any compromise. I want to do a small recap on the topic here.

Let's start from the beginning: Fastify is a CJS module written in JS. We need to use this as a base for its types and implement this as the only way of importing all of its ecosystem.
Typescript provides the syntax for handling this cases that must be used in order to have correct types is`export =`

On the TS consumer side, we have to remember that all of the modules that are consumed in TS but are written in CJS has to be imported like:
`import * as fastify from 'fastify'`
`import fastify from 'fastify'` is only allowed is `esModulesInterop: true` is used inside tsconfig.json. This is not, however, the default behavior and it will never be.

In order to support also ES modules we are adding `.default` prop to our `export = fastify` object (`fastify.default = fastify`). In this way we will support all of the bundlers out there. 

Speaking of `import('fastify')` actually it is not a case or strange object at all: it returns a promise that resolves with whatever is the export of that ES module: here the important part is "ES module". For example:
```
{
    default: typeof import("fastify");
    otherNamedExports: any 
    ...
}
```
As you can see we are covered also for the dynamic import case. 


To recap:
1. we need to update docs regarding TS imports to point out the nature of the CJS fastify's module
2. we need to support default imports if fastify-plugin is used: adding `.default` prop if it is missing looks fine to me
3. we need to encourage plugin autors to use  `export =` syntax and to add `.default` prop

I hope this helps, and if needed i can champion this across the ecosystem.

As a reference you can check https://github.com/fox1t/fastify-multer/blob/master/src/index.ts#L129 even if it is written in TS can and need to mimic this in JS written Fastify ecosystem typings.

**Comment by hongyiweiwu** [comment]
Hey @fox1t thanks for the post! This is probably the best way to go about this right now -- however since you're using `export =`, you can only export the plugin itself. This won't work if the plugin also wants to, say export an options type or some other helper method, which might be necessary. At that point we may have to resort to the scenarios in the post I made above.

I think after combining our solutions we will have the complete solution to the problem and that can be added into the docs. Meanwhile I agree with @mcollina that the cleanest way to go about it is to add the `default` logic into `fastify-plugin`.

**Comment by fox1t** [comment]
I have some more context about dynamic`import()` in TypeScript. It seems that it plays very bad with `esModuleInterop: true` and CJS exported modules (even if they have a .default prop). Even with all the knowledge i have about TS, I can't figure out why it works like this. It doubles `.default` prop and types check doesn't pass!
<img width="642" alt="image" src="https://user-images.githubusercontent.com/1620916/87685472-b801f280-c783-11ea-814d-90be6afda8c8.png">

However if I add an explicit `.default` to the import it works as expected:
<img width="673" alt="image" src="https://user-images.githubusercontent.com/1620916/87685674-f4cde980-c783-11ea-8f41-31c22d8428ed.png">

Setting `esModuleInterop: false` it starts to work also without explicit .default:
<img width="535" alt="image" src="https://user-images.githubusercontent.com/1620916/87685862-25ae1e80-c784-11ea-9eac-914f88e6ba82.png">

We can say that `import()` and `import * as smth from 'smth'` are working differently when using `esModuleInterop: true`.
If I change 
```ts
declare const fastifySwagger: FastifyPluginAsync<SwaggerOptions> & {
  default: FastifyPluginAsync<SwaggerOptions>;
};
```
to 
```ts
declare const fastifySwagger: FastifyPluginAsync<SwaggerOptions>
```
it works both with `import * as swagger from 'fastify-swagger'` and `const swagger = await import('fastify-swagger')`, but not with `import swagger from 'fastify-swagger'` if `esModuleInterop: false`. On the other side if `esModuleInterop: true` it works both with `import swagger from 'fastify-swagger'` and `const swagger = await import('fastify-swagger')` but not 
`import * as swagger from 'fastify-swagger'`.

So here we have to choose how we want to export our types. Saying that in JS is correct to have also `.default` exported to support ESM, we can:
a. make "real types" considering that it is a CJS module and it is correct to use `export =` without adding `.default` to types
b. add `.default` and make a docs clear that if you want to use `import()` and `esModuleInterop: true` you need to add a `.default` after importing it.

**Comment by fox1t** [comment]
> Hey @fox1t thanks for the post! This is probably the best way to go about this right now -- however since you're using `export =`, you can only export the plugin itself. This won't work if the plugin also wants to, say export an options type or some other helper method, which might be necessary. At that point we may have to resort to the scenarios in the post I made above.
> 
> I think after combining our solutions we will have the complete solution to the problem and that can be added into the docs. Meanwhile I agree with @mcollina and you that the cleanest way to go about it is to add the `default` logic into `fastify-plugin`.

Using `export =` doesn't prevent you exporting other things:
<img width="685" alt="image" src="https://user-images.githubusercontent.com/1620916/87687232-ca7d2b80-c785-11ea-8b35-28836896e837.png">
What you can do in exports.module, you can do it also using `export =` syntax
<img width="679" alt="image" src="https://user-images.githubusercontent.com/1620916/87687396-f7314300-c785-11ea-9634-637737722246.png">

**Comment by fox1t** [comment]
Here we need also to ask @Ethan-Arrowood and @delvedor what they think on all of this. 

If needed I can write a doc that sum-ups all of this, since I can understand that it is not easy to follow, but we surely need to make a decision to prevent happening again what is happening after the v3.

**Comment by mcollina** [comment]
I think a write up would be useful. I also don't understand if this was broken in v3 or it was always there.

**Comment by hongyiweiwu** [comment]
> What you can do in exports.module, you can do it also using `export =` syntax

Oh yes, but not anything on TypeScript level, like a type or an interface, if I understand correctly?

**Comment by hongyiweiwu** [comment]
> I have some more context about dynamic`import()` in TypeScript. It seems that it plays very bad with `esModuleInterop: true` and CJS exported modules (even if they have a .default prop). Even with all the knowledge i have about TS, I can't figure out why it works like this. It doubles `.default` prop and types check doesn't pass!
> <img alt="image" width="642" src="https://user-images.githubusercontent.com/1620916/87685472-b801f280-c783-11ea-814d-90be6afda8c8.png">
> 
> However if I add an explicit `.default` to the import it works as expected:
> <img alt="image" width="673" src="https://user-images.githubusercontent.com/1620916/87685674-f4cde980-c783-11ea-8f41-31c22d8428ed.png">
> 
> Setting `esModuleInterop: false` it starts to work also without explicit .default:
> <img alt="image" width="535" src="https://user-images.githubusercontent.com/1620916/87685862-25ae1e80-c784-11ea-9eac-914f88e6ba82.png">
> 
> We can say that `import()` and `import * as smth from 'smth'` are working differently when using `esModuleInterop: true`.
> If I change
> 
> ```ts
> declare const fastifySwagger: FastifyPluginAsync<SwaggerOptions> & {
>   default: FastifyPluginAsync<SwaggerOptions>;
> };
> ```
> 
> to
> 
> ```ts
> declare const fastifySwagger: FastifyPluginAsync<SwaggerOptions>
> ```
> 
> it works both with `import * as swagger from 'fastify-swagger'` and `const swagger = await import('fastify-swagger')`, but not with `import swagger from 'fastify-swagger'` if `esModuleInterop: false`. On the other side if `esModuleInterop: true` it works both with `import swagger from 'fastify-swagger'` and `const swagger = await import('fastify-swagger')` but not
> `import * as swagger from 'fastify-swagger'`.
> 
> So here we have to choose how we want to export our types. Saying that in JS is correct to have also `.default` exported to support ESM, we can:
> a. make "real types" considering that it is a CJS module and it is correct to use `export =` without adding `.default` to types
> b. add `.default` and make a docs clear that if you want to use `import()` and `esModuleInterop: true` you need to add a `.default` after importing it.

This most likely is because when enabling the `esModuleInterop` TypeScript automatically enables `allowSyntheticImport` which automatically adds a default property to your export. You can enable `esModuleInterop` and then disable `allowSyntheticImport` and then it would work.

**Comment by hongyiweiwu** [comment]
> > I have some more context about dynamic`import()` in TypeScript. It seems that it plays very bad with `esModuleInterop: true` and CJS exported modules (even if they have a .default prop). Even with all the knowledge i have about TS, I can't figure out why it works like this. It doubles `.default` prop and types check doesn't pass!
This most likely is because when enabling the `esModuleInterop` TypeScript automatically enables `allowSyntheticImport` which automatically adds a default property to your export. You can enable `esModuleInterop` and then disable `allowSyntheticImport` and then it would work.

**Comment by Ethan-Arrowood** [comment]
A write up would be appreciated. The `esModuleInterop` and `allowSyntheticImport` properties should be included as this is hard to follow. I think all in all we need to figure out a recommended process for** creating a typed fastify plugin** (either written in TS or an accompanying type definition file). Our users should not be expected to understand all of this stuff just to create a plugin for fastify (but for maintainers and contributors this can be extremely helpful information).

**Comment by fox1t** [comment]
> > I have some more context about dynamic`import()` in TypeScript. It seems that it plays very bad with `esModuleInterop: true` and CJS exported modules (even if they have a .default prop). Even with all the knowledge i have about TS, I can't figure out why it works like this. It doubles `.default` prop and types check doesn't pass!
> 
> This most likely is because when enabling the `esModuleInterop` TypeScript automatically enables `allowSyntheticImport` which automatically adds a default property to your export. You can enable `esModuleInterop` and then disable `allowSyntheticImport` and then it would work.

It is not that simple: `esModuleInterop: true` makes work both `import swagger from 'fastify-swagger` and `import * as swagger from 'fastify-swagger`. If what you said was correct also `import swagger from 'fastify-swagger` wouldn't work :)

It is an issue just with import() that doesn't check for `.default` prop before adding it.

**Comment by fox1t** [comment]
> > What you can do in exports.module, you can do it also using `export =` syntax
> 
> Actually scratch that I saw you could use `typeof`. In this case I think this is the best way to go forward.

As I said 100% of `module.exports` functionalities are covered!

**Comment by fox1t** [comment]
> A write up would be appreciated. The `esModuleInterop` and `allowSyntheticImport` properties should be included as this is hard to follow. I think all in all we need to figure out a recommended process for** creating a typed fastify plugin** (either written in TS or an accompanying type definition file). Our users should not be expected to understand all of this stuff just to create a plugin for fastify (but for maintainers and contributors this can be extremely helpful information).

I agree 100% with you. Here i am trying to find a consensus on what we consider to be the correct implementation and support regarding CJS, ESM and TS. Unfortunately this never were stable in the Fastify's ecosystem and I think that now we are mature enough as project to fix all of these issues. :)

**Comment by fox1t** [comment]
TS `import()` "bug" reference: As supposed it acts differently from the ES2020 one here and I think that, since typeof is used here, creates also types our mismatch: https://github.com/microsoft/TypeScript/issues/27678

Here is another reference that explains non-consistent behavior of the current implementation: https://mariusschulz.com/blog/dynamic-import-expressions-in-typescript

**Comment by fox1t** [comment]
> > What you can do in exports.module, you can do it also using `export =` syntax
> 
> Actually scratch that I saw you could use `typeof`. In this case I think this is the best way to go forward.

I've just discovered that even if you can export a typeof something when you try to import it, TS considers it as executable code and not just type alias: that's it, it thinks that there is a bar variable. So you were right here, we are not going to use `export =`. (this is just a tease of a document I am preparing, but kudos to you @hongyiweiwu :) )

**Comment by hongyiweiwu** [comment]
> I've just discovered that even if you can export a typeof something when you try to import it, TS considers it as executable code and not just type alias: that's it, it thinks that there is a bar variable. So you were right here, we are not going to use `export =`. (this is just a tease of a document I am preparing, but kudos to you @hongyiweiwu :) )

Glad to be of help! In this way what do you think you'll be using to export this stuff?

**Comment by fox1t** [comment]
Check this repository for a possibile solution: https://github.com/fox1t/modules-playground
